hugo/content/docs/model_sources.md

```diff
@@ -0,0 +1,146 @@
+
+On client-side, instead of registering an actual `ModelSource` we use a `DiagramServerProxy`. The Proxy handles the communication and forwards actions to the `ActionDipatcher`. Out of the box, Sprotty offers the `WebSocketDiagramServerProxy` for communicating through WebSockets with the `DiagramServer`.
+Should a different form of communication be necessary we would have to [create a custom proxy](#creating-a-custom-model-source-proxy).
+
+Using the `WebSocketDiagramServerProxy` is simple. we just need to call `listen` on the ModelSource and hand it the WebSocket we're communicating with.
+```Typescript
+const modelSource = container.get<WebSocketDiagramServerProxy>(TYPES.ModelSource);
+modelSource.listen(websocket);
+```
+
+For creating `DiagramServer` itself, let's look at an example. 
+```Typescript
+// Creating a new websocket server
+const wss = new WebSocketServer.Server({ port: 8080 });
+
+// create our DiagramServices
+const elkFactory: ElkFactory = () => new SocketElkServer();
+const services: DiagramServices = {
+    DiagramGenerator: new RandomGraphGenerator(),
+    ModelLayoutEngine: new ElkLayoutEngine(elkFactory)
+}
+ 
+// Creating connection using websocket
+wss.on("connection", ws => {
+
+    const diagramServer = new DiagramServer(action => {
+        ws.send(JSON.stringify(action));
+    }, services)
+
+    ws.on('message' data => {
+        diagramServer.accept(data.action);
+    });
+});
+```
+In this, we assume we have a simple [nodeJs WebSocket server](https://github.com/websockets/ws) and want to create a `DiagramServer` for it.
+
+As we can see, there are two parts to creating the `DiagramServer`. 
+First, we need a dispatch method to send actions from the server to the client. This can be as simple as calling `ws.send()` with the serialized action.
+Second, we need the `DiagramServices`. the `DiagramServices` type looks like this:
+```Typescript
+export interface DiagramServices {
+    readonly DiagramGenerator: IDiagramGenerator
+    readonly ModelLayoutEngine?: IModelLayoutEngine
+    readonly ServerActionHandlerRegistry?: ServerActionHandlerRegistry
+}
+```
+There are 3 components to the `DiagramServices`. One is mandatory, the other ones optional:
+- The `DiagramGenerator` which the server uses to create the schema of the Diagram. 
+- Optionally the `ModelLayoutEngine`, like the `ElkLayoutEngine` from [sprotty-elk](https://github.com/eclipse-sprotty/sprotty/tree/master/packages/sprotty-elk), if we want to do server-side layouting.
+- Optionally the `ServerActionHandlerRegistry` for overwriting the default handling of incoming actions. 
+
+## Creating a custom model source proxy
+In case communication between the `DiagramServer` and client does not work through WebSockets, for example when the `DiagramServer` is running in a worker or the sprotty client is in a vscode webview (see [sprotty-vscode](https://github.com/eclipse-sprotty/sprotty-vscode)), we can easily implement our own proxy.
```

Comment:
```suggestion
In case communication between the `DiagramServer` and client does not work through WebSockets, for example when the `DiagramServer` is running in a worker or the sprotty client is in a vscode webview (see [sprotty-vscode](https://github.com/eclipse-sprotty/sprotty-vscode)), we can easily implement our own proxy instead.
```